app/src/main/java/com/selfdrivingboat/boatcontroller/SelfDrivingBoatCamera.java

```diff
@@ -0,0 +1,268 @@
+                    output.close();
+                }
+            }
+        }
+    };
+
+
+    /*
+     This is the callback necessary for the manager.openCamera Call back needed above.
+    */
+    // the callback says "when the camera is opened" give me the camera and put it on mCameraDevice
+    private CameraDevice.StateCallback mStateCallback = new CameraDevice.StateCallback() {
+
+        @Override
+        public void onOpened(CameraDevice camera) {
+            mCameraDevice = camera;
+            //setup the capture of the current surface.
+            //is the setup to take the picture, now the mCameraDevice is initialized.
+            //configure the catureBuilder, which is built in listener later on.
+            try {
+                captureBuilder = mCameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE);
+            } catch (CameraAccessException e) {
+                e.printStackTrace();
+            }
+
+            captureBuilder.addTarget(reader.getSurface());
+            captureBuilder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO);
+        }
+
+
+        @Override
+        public void onDisconnected(CameraDevice camera) {
+        }
+
+        @Override
+        public void onError(CameraDevice camera, int error) {
+        }
+
+    };
+
+    // alex: openCamera basically set up the camera with the callback
+    private void openCamera() {
+
+        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
+            try {
+                String cameraId = manager.getCameraIdList()[0];
+
+                // setup the camera perview.  should wrap this in a checkpermissions, which studio is bitching about
+                // except it has been done before this fragment is called.
+                manager.openCamera(cameraId, mStateCallback, null);
+
+            } catch (CameraAccessException e) {
+                e.printStackTrace();
+            }
+        } else {
+            // log something
+        }
+    }
+
+    CameraCaptureSession.CaptureCallback captureListener = new CameraCaptureSession.CaptureCallback() {
+
+        @Override
+        public void onCaptureCompleted(CameraCaptureSession session,
+                                       CaptureRequest request, TotalCaptureResult result) {
+
+            super.onCaptureCompleted(session, request, result);
+            //new tell the system that the file exists , so it will show up in gallery/photos/whatever
+            ContentValues values = new ContentValues();
+            values.put(MediaStore.Images.Media.DATE_TAKEN, System.currentTimeMillis());
+            values.put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg");
+            values.put(MediaStore.MediaColumns.DATA, file.toString());
+            context.getContentResolver().insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values);
+
+            // alex: here we saved the image
```

Comment:
adding the logging looks like

Log.i("we saved the image at ... ")

do it everywhere in the code and look in the console at your logs. This is both to keep track what the code is doing during execution and also to make sure you understand the code